Foundation/NSPathUtilities.swift

```diff
@@ -10,13 +10,22 @@
 import CoreFoundation
 
 public func NSTemporaryDirectory() -> String {
-    #if os(macOS) || os(iOS)
-    var buf = [Int8](repeating: 0, count: 100)
-    let r = confstr(_CS_DARWIN_USER_TEMP_DIR, &buf, buf.count)
-    if r != 0 && r < buf.count {
-        return String(cString: buf, encoding: .utf8)!
-    }
-    #endif
+#if os(macOS) || os(iOS)
+    var length: Int = confstr(_CS_DARWIN_USER_TEMP_DIR, nil, 0)
+    guard length > 0 else { return "/tmp" }
+
+    var buffer: [Int8] = Array<Int8>(repeating: 0, count: length)
+    length = confstr(_CS_DARWIN_USER_TEMP_DIR, &buffer, buffer.count)
+    precondition(length > 0 && length < buffer.count,
+                 "_CS_DARWIN_USER_TEMP_DIR mutation race")
+    return String(cString: buffer, encoding: .utf8)!
+#elseif os(Windows)
+    var cchLength: DWORD = GetTempPathW(0, nil)
+    var wszPath: UnsafeMutableBufferPointer = UnsafeMutableBufferPointer<WCHAR>.allocate(capacity: Int(cchLength + 1))
+    defer { wszPath.deallocate() }
+    _ = GetTempPathW(cchLength + 1, wszPath.baseAddress)
```

Comment:
The docs say this function can fail. You should check the return value and fall back to the env variable path.

Reply:
https://docs.microsoft.com/en-us/windows/desktop/api/fileapi/nf-fileapi-gettemppathw

Reply:
We just invoked it earlier and checked the result (to determine the buffer size to ensure that we have the properly sized buffer.  The second invocation has already gone through the environment and failing that has returned `%SystemRoot%\Temp` (aka `C:\Windows\Temp`).

Reply:
TOCTOU race - the environment may have been changed by another thread between the two calls.

Reply:
Yeah, I have an assertion in the Darwin path for exactly that.  I suppose that I can add a second check here too.  But, that is going into #1940.